binary_search_tree/tree.py

```diff
@@ -7,52 +7,129 @@ def __init__(self, key, val = None):
         self.value = val
         self.left = None
         self.right = None
-        
-
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    def add_helper(self, current_node, new_node):
+        if current_node == None:
+            return new_node
+
+        if new_node.key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, new_node)
+        else:
+            current_node.right = self.add_helper(current_node.right, new_node)
+
+        return current_node
+
+    # # Time Complexity: O(log n)
+    # # Space Complexity: O(1)
+    def add(self, key, value=None):
+        node = TreeNode(key, value)
+        if self.root == None:
+            self.root = node
+            return
+
+        self.add_helper(self.root, node)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def find_helper(self, current, key):
+        if current == None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif key < current.key:
+            return self.find_helper(current.left, key)
+
+        return self.find_helper(current.right, key)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
```

Comment:
👍 Since you're doing this recursively the space complexity is O(log n) because of the call stack.